packages/vscode-messenger-common/src/messages.ts

```diff
@@ -176,8 +176,83 @@ export type NotificationHandler<P> = (params: P, sender: MessageParticipant) =>
  * Base API for Messenger implementations.
  */
 export interface MessengerAPI {
-    sendRequest<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params?: P): Promise<R>
+    sendRequest<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params?: P, cancelable?: Cancelable): Promise<R>
     onRequest<P, R>(type: RequestType<P, R>, handler: RequestHandler<P, R>): void
     sendNotification<P>(type: NotificationType<P>, receiver: MessageParticipant, params?: P): void
     onNotification<P>(type: NotificationType<P>, handler: NotificationHandler<P>): void
 }
+
+/**
+ *  Deferred promise that can be resolved or rejected later.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export class PendingRequest<R = any> {
+    resolve: (value: R) => void;
+    reject: (reason?: unknown) => void;
+
+    result = new Promise<R>((resolve, reject) => {
+        this.resolve = (arg) => resolve(arg);
+        this.reject = (err) => reject(err);
+    });
+}
+
+/**
+ * Interface that allows to check for cancellation and
+ * set a listener that is called when the request is canceled.
+ */
+export interface CancelIndicator {
+    isCanceled(): boolean;
```

Comment:
This should be a readonly property, not a function (it can be implemented with a getter).